### PR TITLE
Log command

### LINKED ISSUE #278: [2pt] We need `cartridge log` command
Since `cartridge start -d` redirects logs to a file, we can reveal instance(s) logs using `tail -f`. 

I suggest to implement `cartridge log` command:

`cartridge log [INSTANCE_NAME...] [flags]` show instance(s) logs. 
If no ` [INSTANCE_NAME...]` are specified, instances are discovered from cfg file (just like `start/stop/status` do).

Options:

* `-f, --follow` output appended data as the file grows;
* `-n, --lines=NUM` output the last NUM lines.

I think, it would be good to show logs with colorized prefixes (just like start in interactive mode does).

### PR DESCRIPTION
**Pull request by dokshina**
Closes #278

### CODE REVIEW

File: cli/commands/log.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+	logCmd.Flags().BoolVar(&projectCtx.StateboardOnly, "stateboard-only", false, stateboardOnlyFlagDoc)
+
+	logCmd.Flags().BoolVarP(&followLog, "follow", "f", false, followFlagDoc)
+	logCmd.Flags().IntVarP(&linesLog, "lines", "n", defaultLinesLog, followFlagDoc)
+}
+
+var logCmd = &cobra.Command{
+	Use:   "log [INSTANCE_NAME...]",
+	Short: "Get logs of instance(s)",
+	Long:  fmt.Sprintf("Get logs of instance(s)n\n%s", runningCommonDoc),
```
Comment (olegrok) [inline cli/commands/log.go:43]: `instance(s)n` -> `instance(s)`?

File: cli/common/utils.go [hunk 2]
```diff
@@ -143,3 +145,91 @@ func ParseYmlFile(path string) (map[string]interface{}, error) {
...
+	buf := make([]byte, bufSize)
+
+	var filePos int64 = fileSize - bufSize
+	var lastNewLinePos int64 = 0
+	var newLinesN int = 0
+
+	// check last symbol of the last line
+
+	if err := readFromPos(f, fileSize-1, &buf); err != nil {
+		return 0, fmt.Errorf("%s", err)
```
Comment (olegrok) [inline cli/common/utils.go:193]: `return 0, err`

File: cli/common/utils.go [hunk 3]
```diff
@@ -143,3 +145,91 @@ func ParseYmlFile(path string) (map[string]interface{}, error) {
...
+	}
+	if buf[0] != '\n' {
+		newLinesN++
+	}
+
+	lastPart := false
+
+Loop:
+	for {
+		if filePos < 0 {
```
Comment (olegrok) [inline cli/common/utils.go:203]: Is it possible?

File: cli/common/utils.go [hunk 4]
```diff
@@ -143,3 +145,91 @@ func ParseYmlFile(path string) (map[string]interface{}, error) {
...
+			filePos = 0
+			lastPart = true
+
+			buf = make([]byte, fileSize%bufSize)
+		}
+
+		if err := readFromPos(f, filePos, &buf); err != nil {
+			return 0, fmt.Errorf("%s", err)
```
Comment (olegrok) [inline cli/common/utils.go:211]: `return 0, err`

File: cli/common/utils.go [hunk 5]
```diff
@@ -143,3 +145,91 @@ func ParseYmlFile(path string) (map[string]interface{}, error) {
...
+		}
+
+		for i := len(buf) - 1; i >= 0; i-- {
```
Comment (olegrok) [inline cli/common/utils.go:214]: Seems it could be moved into separate function

File: cli/running/processes_set.go [hunk 6]
```diff
@@ -231,3 +231,61 @@ func (set *ProcessesSet) Status() error {
...
+		time.Sleep(100 * time.Millisecond)
+	}
+
+	var errors []error
+
+	// wait for all processes result
+	for i := 0; i < len(*set); i++ {
+		select {
+		case res := <-resCh:
+			log.Infof(getResStr(&res))
```
Comment (olegrok) [inline cli/running/processes_set.go:272]: Is there simple `log.Info` without `f`?

File: test/project.py [hunk 7]
```diff
@@ -255,3 +252,51 @@ def patch_init_to_send_statuses(project, statuses):
 
     with open(os.path.join(project.path, 'stateboard.init.lua'), 'w') as f:
         f.write(patched_init)
+
+
+def patch_init_to_log_lines(project, lines):
```
Comment (olegrok) [inline test/project.py:257]: Write simple description for it.

File: cli/running/processes_set.go (hunk 6 above)
Comment (dokshina) [reply cli/running/processes_set.go:272]: Yes, in this case function just returns.

File: cli/common/utils.go (hunk 5 above)
Comment (dokshina) [reply cli/common/utils.go:214]: This is the main part of this function, I'm not sure that is should be separated.

File: test/project.py (hunk 7 above)
Comment (dokshina) [reply test/project.py:265]: Added description.

File: cli/commands/log.go [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+	if err := running.Log(&projectCtx, followLog, linesLog); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+const (
+	followFlagDoc = `Output appended data as the log grows
```
Comment (olegrok) [inline cli/commands/log.go:75]: `follow` description from `man journalctl`.
```
Show only the most recent journal entries, and continuously print new entries as they are appended to the journal.
```

File: cli/common/utils.go [hunk 9]
```diff
@@ -16,6 +16,8 @@ import (
 	"gopkg.in/yaml.v2"
 )
 
+var bufSize int64 = 10000
```
Comment (olegrok) [inline cli/common/utils.go:19]: Why is it not const?

File: cli/common/utils.go [hunk 10]
```diff
@@ -143,3 +145,95 @@ func ParseYmlFile(path string) (map[string]interface{}, error) {
...
+		return 0, nil
+	}
+
+	buf := make([]byte, bufSize)
+
+	var filePos int64 = fileSize - bufSize
+	var lastNewLinePos int64 = 0
+	var newLinesN int = 0
+
+	// check last symbol of the last line
```
Comment (olegrok) [inline cli/common/utils.go:194]: Why is this case separated from others?

File: cli/common/utils.go [hunk 11]
```diff
@@ -143,3 +145,95 @@ func ParseYmlFile(path string) (map[string]interface{}, error) {
...
+
+	lastPart := false
+
+Loop:
+	for {
+		if filePos < 0 {
+			filePos = 0
+			lastPart = true
+
+			buf = make([]byte, fileSize%bufSize)
```
Comment (olegrok) [inline cli/common/utils.go:211]: I suggest you could avoid recreation of buffer for each iteration. Create it once and reuse.

File: cli/common/utils.go [hunk 12]
```diff
@@ -143,3 +145,95 @@ func ParseYmlFile(path string) (map[string]interface{}, error) {
...
+
+	if _, err := f.Read(*buf); err != nil {
+		return fmt.Errorf("Failed to read: %s", err)
+	}
+
+	return nil
+}
+
+// GetLastNLinesBegin return the position of last n lines begin
+func GetLastNLinesBegin(filepath string, n int) (int64, error) {
```
Comment (olegrok) [inline cli/common/utils.go:162]: BTW, have you measure performance of this implementation on big files? Say, 1GB and more.

File: cli/common/utils.go (hunk 9 above)
Comment (dokshina) [reply cli/common/utils.go:19]: To be changed in test.

File: cli/common/utils.go (hunk 10 above)
Comment (dokshina) [reply cli/common/utils.go:194]: To process `one\ntwo` and `one\ntwo\n` cases the same way.

File: cli/common/utils.go (hunk 11 above)
Comment (dokshina) [reply cli/common/utils.go:211]: Buffer is recreated once at the last part. It's size can be less than `bufSize` and `len(buf)` returns `bufSize`, no matter how many bytes was read.

File: cli/commands/log.go (hunk 8 above)
Comment (dokshina) [reply cli/commands/log.go:71]: I was thinking about `tail` command writing this feature. Why do you think that current description is bad?

Comment (olegrok) [reply cli/commands/log.go:69]: > Why do you think that current description is bad?

I don't. But I suggest you an alternative.

File: cli/common/utils.go (hunk 11 above)
Comment (dokshina) [reply cli/common/utils.go]: I've removed the buffer recreation.

**Pull request merged**